services/QuillLMS/db/migrate/20251029210337_convert_examples_to_sti.evidence.rb

```diff
@@ -0,0 +1,370 @@
+      t.string :curriculum_assigned_status, null: false
+      t.string :curriculum_label
+      t.text :curriculum_proposed_feedback
+      t.text :highlight
+      t.text :notes
+      t.text :student_response, null: false
+      t.string :type, null: false
+      t.integer :cluster_id
+      t.integer :dataset_id, null: false
+      t.string :original_type
+      t.integer :original_id
+      t.timestamps
+    end
+
+    add_index :evidence_research_gen_ai_examples, :type
+    add_index :evidence_research_gen_ai_examples, [:original_type, :original_id]
+
+    verify_no_unexpected_foreign_keys
+    migrate_prompt_examples
+    migrate_test_examples
+    migrate_draft_examples
+    reset_sequence
+    update_foreign_keys
+    verify_migration
+  end
+
+  private def verify_no_unexpected_foreign_keys
+    Rails.logger.debug 'Checking for unexpected foreign key references...'
+
+    known_fk_tables = %w[
+      evidence_research_gen_ai_llm_prompt_prompt_examples
+      evidence_research_gen_ai_llm_examples
+      evidence_research_gen_ai_workflow_executions
+    ]
+
+    fk_mappings = {
+      prompt_example_id: 'evidence_research_gen_ai_prompt_examples',
+      test_example_id: 'evidence_research_gen_ai_test_examples',
+      draft_example_id: 'evidence_research_gen_ai_draft_examples'
+    }
+
+    unexpected_references = find_unexpected_references(known_fk_tables, fk_mappings)
+
+    raise_if_unexpected_references_found(unexpected_references)
+
+    Rails.logger.debug '✓ No unexpected foreign key references found'
+  end
+
+  private def find_unexpected_references(known_tables, fk_mappings)
+    unexpected = []
+    all_tables = ActiveRecord::Base.connection.tables
+
+    all_tables.each do |table_name|
+      next if known_tables.include?(table_name)
+
+      fk_mappings.each do |fk_column, referenced_table|
+        unexpected.concat(check_table_for_fk(table_name, fk_column, referenced_table))
+      end
+    end
+
+    unexpected
+  end
+
+  private def check_table_for_fk(table_name, fk_column, referenced_table)
+    return [] unless ActiveRecord::Base.connection.column_exists?(table_name, fk_column)
+
+    count = check_foreign_key_count(table_name, fk_column, referenced_table)
+    return [] if count.zero?
+
+    ["#{table_name}.#{fk_column} (#{count} rows)"]
+  end
+
+  private def check_foreign_key_count(referencing_table, fk_column, referenced_table)
+    result = execute(<<-SQL.squish)
+      SELECT COUNT(*)
+      FROM #{referencing_table}
+      WHERE #{fk_column} IN (SELECT id FROM #{referenced_table})
+    SQL
+    result.getvalue(0, 0).to_i
+  end
+
+  private def raise_if_unexpected_references_found(unexpected_references)
+    return if unexpected_references.empty?
+
+    error_msg = "Found unexpected foreign key references that need to be migrated:\n  - #{unexpected_references.join("\n  - ")}"
+    Rails.logger.error error_msg
+    raise error_msg
+  end
+
+  private def migrate_prompt_examples
```

Comment:
Also note, I think looping through the records is preferred, but if you still end up hitting performance issues, as a secondary option we could do this as a mass insert (or upsert) into the new tables.